### ISSUE TITLE
Add "with resources" feature to Try

### ISSUE THREAD

**Issue by Jezorko** [Jezorko/kofun#19]
# Case study: TryWithResources

## case: reading resource once and then closing it

  Pros:
   * very simple implementation
   * transparent resource scope (if methods named correctly and JavaDoc provided - I'm looking at you vavr)
   
  Cons:
   * limited usage
   * no flexibility

### Fluent example:

```java
Try.withResources(Resource::new)
   .of(Resource::readData) // resource is closed
   ... // continue with Try of data
```

### Plain Java example:

```java
try (Resource resource = new Resource()) {
    Data data = resource.readData();
    ...
}
```

## case: using the resource in transformations

  Pros:
   * transparent resource scope
   * better control over the logic
   * flat call structure
   
  Cons:
   * rather difficult to implement
   * explicit resource scope declaration
   * scope is not forced (logical leaks are possible)

### Fluent example:

```java
Try.withResources(Connection::new)
   .map(Connection::readMessage) // first call gives us a pair of values
   .map((connection, message) -> transform(connection.getAddress(), message)) // resource still alive so we can read more attributes
   .onSuccess(Connection::writeMessage)
   .cleanUp() // resource is closed
   .onError(logger::log);
```
### Plain Java example:

```java
try (Connection connection = new Connection()) {
    Message message = connection.readMessage();
    Message reply = transformData(message);
    connection.writeMessage(reply);
} catch (Exception error) {
    logger.log(error);
}
```

## case: reading resource in a loop

  Pros:
   * no need to repeatedly open and close the resource
   * no need to put everything that uses the resource in a single method / lambda
   * can use fluent semantics
   * loopWhile could have more variants to provide better flexibility
   
  Cons:
   * very difficult to implement
   * similar to the Observer pattern (out of scope)
   * ambiguous resource scope
   
### Fluent example:

```java
Try.withResources(Connection::new)
   .loopWhile(Connection::hasNewMessages)
   .map(Connection::readMessage)
   .map(this::transformData)
   .onSuccess(Connection::writeMessage)
   .onError(logger:log);
```

### Plain Java example:

```java
try (Connection connection = new Connection()) {
    while (connection.hasNewMessages()) {
        try {
            Message message = connection.readMessage();
            Message reply = transformData(message);
        } catch (Exception error) {
            logger.log(error);
        }
    }
}
```

**Comment by Jezorko** [comment]
Blocked by #12
